### PR TITLE
add RNTester-E2E: tests for iOS and Android via Appium, WDIO and Jest

### PR DESCRIPTION
**Pull request by kelset**
## Summary:

The motivation is to create an E2E testing solution for the RNTester app that can be used to flag when things break and the release crew can use to validate more of the codebase ahead of a release.

This first PR wants to just showcase the main flow (how tests are made, where the E2E folder lives, how it's used by CircleCI) and then we can build on top of it with subsequent PRs, with the help of an umbrella issue to get the community involved in making more tests!

This work is being done cooperation with Callstack developers (@mateuszm22 , @adzironman, and others) - reason why the branch [lives in a fork](https://github.com/mateuszm22/react-native/tree/k%2Bm/new-rn-tester-E2E).

### Extra notes

We are still on WebDriverIO 7 because during the exploratory phase, WDIO was a bit problematic and the workarounds needed to make it work don't seem very nice - see this PR: https://github.com/kelset/react-native-e2e-jest-appium-webdriverio/pull/4 --- this will be revisited in the future, once there's a better path for upgrading.

### Things that will be handled as follow up work

* upgrade to WDIO 8
* [an automated yarn run-e2e-test script](https://github.com/facebook/react-native/pull/36267#discussion_r1269378065)
* [a small refactoring into a js script](https://github.com/facebook/react-native/pull/36267#discussion_r1272050735)

## Changelog:

[INTERNAL] [ADDED] - Added first working configuration for e2e testing

## Test Plan:

You can play with this locally by [follow the readme](https://github.com/mateuszm22/react-native/blob/k+m/new-rn-tester-E2E/packages/rn-tester-e2e/README.md).

CircleCI jobs will be added.

### CODE REVIEW

File: packages/rn-tester-e2e/test/screenObjects/buttonComponent.screen.js
Comment (kelset) [inline packages/rn-tester-e2e/test/screenObjects/buttonComponent.screen.js:21]: we need to change things to be like here https://github.com/kelset/react-native-e2e-jest-appium-webdriverio/blob/main/__tests__/basicE2E.test.js#L33

File: packages/rn-tester-e2e/README.md
Comment (szymonrybczak) [inline packages/rn-tester-e2e/README.md:94]:
```suggestion
And modify lines L24->L39 to reflect your local setup configuration (ex. `platformVersion`, `deviceName`).
```

File: packages/rn-tester-e2e/README.md
Comment (szymonrybczak) [inline packages/rn-tester-e2e/README.md:75]: I think we shouldn't use here hardcoded architecture `x86_64`, because this may be different for users based on their setup.

Comment (kelset) [reply packages/rn-tester-e2e/README.md:75]:
```suggestion
4. copy the app `app-*-debug.apk` to the following directory `/react-native/packages/rn-tester-e2e/apps`
```

File: packages/rn-tester-e2e/tests/screens/components.screen.js
Comment (cipolleschi) [inline packages/rn-tester-e2e/tests/screens/components.screen.js:19]: Can we extract the `Button Simple React Native button component.` string in a variable? if something changes, it will be easier to keep it in synch

File: packages/rn-tester-e2e/tests/screens/components/buttonComponent.screen.js
Comment (cipolleschi) [inline packages/rn-tester-e2e/tests/screens/components/buttonComponent.screen.js:16]: I think we can have better API's here... 
like
```
// in Utils

function iOSLabel(label: string): string {
  return `[label="${label}"]`;
}

// Used like this.
Utils.iOSLabel("Button")
```

File: packages/rn-tester-e2e/tests/screens/components/buttonComponent.screen.js
Comment (cipolleschi) [inline packages/rn-tester-e2e/tests/screens/components/buttonComponent.screen.js:23]: Similar to above, we can have something like

```
function androidWidget(type: string, selector: string, id: string): string {
  return `//android.widget.${type}[@${selector}="${id}"]`;
}
```

File: packages/rn-tester-e2e/tests/specs/components/buttonComponentScreen.test.js
Comment (cipolleschi) [inline packages/rn-tester-e2e/tests/specs/components/buttonComponentScreen.test.js:34]: I love these! 😄 ❤️

File: packages/rn-tester-e2e/README.md
Comment (cipolleschi) [inline packages/rn-tester-e2e/README.md:37]:
```suggestion
    yarn test-e2e-local-clean && yarn
```
Just to make sure that the dependencies are reinstalled! ;)

File: packages/rn-tester-e2e/README.md
Comment (cipolleschi) [inline packages/rn-tester-e2e/README.md:56]: I think that we can explore how to tell `xcodebuild` to use a specific derivedData folder. At that point, we can automate these steps.

File: packages/rn-tester-e2e/README.md
Comment (cipolleschi) [inline packages/rn-tester-e2e/README.md:83]: Out of ignorance, but if the e2e folder is in a workspace, shouldn't `yarn` run in the `react-native` root folder also install all the dependencies in the `rn-tester-e2e` folder? 🤔

File: packages/rn-tester-e2e/README.md
Comment (cipolleschi) [inline packages/rn-tester-e2e/README.md:94]: Ideally, we should find a way to run tests on both platforms at the same time... 🤔

**Review by cipolleschi — COMMENTED** [review]
I left some comments that could be addressed, but the work here is super promising! I'm so so excited about this!

File: packages/rn-tester-e2e/README.md
Comment (szymonrybczak) [reply packages/rn-tester-e2e/README.md:56]: There's `-derivedDataPath <path>` option for that (we're using this one internally in React Native CLI).

Comment (cipolleschi) [reply packages/rn-tester-e2e/README.md:56]: yeah, I was referring to that. With that parameter, we can automate all these steps and move the build artifact in the right folder, simplifying the setup (and the CI, probably)

File: packages/rn-tester-e2e/README.md
Comment (NickGerleman) [inline packages/rn-tester-e2e/README.md:21]: Why do we need to do a global install of this dependency, version using the version provided as a dependency?

File: packages/rn-tester-e2e/package.json
Comment (NickGerleman) [inline packages/rn-tester-e2e/package.json:30]: nit: `@types` is typically a devDependency

File: packages/rn-tester-e2e/tests/screens/components.screen.js
Comment (NickGerleman) [inline packages/rn-tester-e2e/tests/screens/components.screen.js:34]: nit: If we are not keeping any state on these classes, maybe we could keep them as a plain object with constants/functions? It would avoid the need to rename these on import.

File: packages/rn-tester-e2e/tests/screens/components.screen.js
Comment (NickGerleman) [inline packages/rn-tester-e2e/tests/screens/components.screen.js:29]:
```suggestion
  async clickButtonComponent(): void {
```

Comment (NickGerleman) [reply packages/rn-tester-e2e/tests/screens/components.screen.js:29]: Same for all of the `Promise<any>` examples below.

File: yarn.lock
Comment (NickGerleman) [inline yarn.lock:13]: > as you quickly should notice, this new folder that is being added packages/rn-tester-e2e at the moment is not part of the yarn workspace because the dependencies it pulls in are quite heavily, so we don't want to force everyone to download them every time. We will have to revisit this once we wire it up on CI

Why are we seeing the dependencies in the root lockfile. Is this change stale?

File: packages/rn-tester-e2e/jest.config.js
Comment (NickGerleman) [inline packages/rn-tester-e2e/jest.config.js:11]: We will also want `maxWorkers: 0` as Jest will normally spawn separate workers in parallel per suite.

File: packages/rn-tester-e2e/jest.config.js
Comment (NickGerleman) [inline packages/rn-tester-e2e/jest.config.js:14]: > A list of paths to modules that run some code to configure or set up the testing framework before each test file in the suite is executed.

Are we wanting to teardown/setup the app for every spec, or do we want to do this once?

**Comment by kelset** [comment]
(adding a quick side-comment here to cross-reference, we're doing something similar in RNTA https://github.com/microsoft/react-native-test-app/pull/1467/files )

File: packages/rn-tester-e2e/jest.config.js
Comment (ThisTallDude) [reply packages/rn-tester-e2e/jest.config.js:14]: Hi @NickGerleman,

We do not want to have flakiness on a high level. We want to maintain a high level of quality and truthfulness.

Currently, we leave these steps before each test as it is.

File: yarn.lock
Comment (mateuszm22) [reply yarn.lock:13]: @kelset WDYT?

File: packages/rn-tester-e2e/jest.config.js
Comment (ThisTallDude) [reply packages/rn-tester-e2e/jest.config.js:11]: Hi @kelset, 
This is fixed in the last MR. Could you please resolve this comment?

File: packages/rn-tester-e2e/tests/screens/components.screen.js
Comment (ThisTallDude) [reply packages/rn-tester-e2e/tests/screens/components.screen.js]: Hi @kelset, 
This is fixed in the last MR. Could you please resolve this comment?

File: packages/rn-tester-e2e/tests/screens/components.screen.js
Comment (ThisTallDude) [reply packages/rn-tester-e2e/tests/screens/components.screen.js]: Hi @kelset, 
This is fixed in the last MR. Could you please resolve this comment?

File: packages/rn-tester-e2e/package.json
Comment (ThisTallDude) [reply packages/rn-tester-e2e/package.json:28]: Hi @kelset, 
This is fixed in the last MR. Could you please resolve this comment?

File: packages/rn-tester-e2e/README.md
Comment (mateuszm22) [reply packages/rn-tester-e2e/README.md:21]: @NickGerleman Yep it's needed to run the instance of appium server.

File: packages/rn-tester-e2e/README.md
Comment (ThisTallDude) [reply packages/rn-tester-e2e/README.md:94]: Hi @cipolleschi @kelset,

We checked that implementing this approach will take some time. We want to create the next PR with this as the next step.

File: packages/rn-tester-e2e/README.md
Comment (ThisTallDude) [reply packages/rn-tester-e2e/README.md:56]: Hi @kelset,
This is fixed in the last MR. Could you please resolve this comment?

File: packages/rn-tester-e2e/README.md
Comment (ThisTallDude) [reply packages/rn-tester-e2e/README.md]: Hi @kelset,
This is fixed in the last MR. Could you please resolve this comment?

File: packages/rn-tester-e2e/tests/screens/components/buttonComponent.screen.js
Comment (ThisTallDude) [reply packages/rn-tester-e2e/tests/screens/components/buttonComponent.screen.js]: Hi @kelset,
This is fixed in the last MR. Could you please resolve this comment?

File: packages/rn-tester-e2e/tests/screens/components/buttonComponent.screen.js
Comment (ThisTallDude) [reply packages/rn-tester-e2e/tests/screens/components/buttonComponent.screen.js]: Hi @kelset,
This is fixed in the last MR. Could you please resolve this comment?

File: packages/rn-tester-e2e/tests/screens/components.screen.js
Comment (ThisTallDude) [reply packages/rn-tester-e2e/tests/screens/components.screen.js]: Hi @kelset,
This is fixed in the last MR. Could you please resolve this comment?

File: packages/rn-tester-e2e/README.md
Comment (cipolleschi) [reply packages/rn-tester-e2e/README.md:56]: What do you mean: `This is fixed in the last MR`? I can still see that the instruction do not explain how to build from command line, passing the `--derivedDataPath`, so it does not seem fixed.

File: yarn.lock
Comment (kelset) [reply yarn.lock]: update on this: we'll actually change approach and add a new CircleCI job that runs this E2E in this PR already.

File: packages/rn-tester-e2e/jest.config.js
Comment (kelset) [reply packages/rn-tester-e2e/jest.config.js:14]: > teardown/setup the app for every spec

yeah that's the better way to reduce flakyness.

**Comment by kelset** [comment]
👋 everyone.

Thank you all for your work and feedback on this PR! I did a meeting yesterday with @cortinico and @cipolleschi and we defined a new plan to ensure that this can land during H2 24; here are the key takeaways:

1) The main change is that we're now aiming to have already this first PR add a CircleCI job that runs the E2E tests on CI.

2) Aside from that, we'll also need to prepare an RFC to back this approach and give a broader vision for how we're aiming at introducing this new level of testing to improve the robustness of the codebase.

We can discuss more details offline, but that's the rough plan and these two things mentioned above are the next steps to make this land.

I've also given a quick pass at the code, rebased on top of main and some small cleanup (and bumping appium, they are entering RC phase so probably 2.0.0 is close).

File: .circleci/config.yml
Comment (cipolleschi) [inline .circleci/config.yml:805]: if the test-e2e produces some artifacts, we can think about saving them in CircleCI using the [`store_artifacts`](https://circleci.com/docs/artifacts/#uploading-artifacts) instruction

**Comment by hurali97** [comment]
Hey folks,

I was reading the RFC and since there's plan to use device farm in the near future, until then I suggest implementing [linkedIn's test butler](https://github.com/linkedin/test-butler) to avoid silly emulator issues in CI. 

However this comes with a constraint that we'll need to use stock android emulator. Now I am not aware of how it will fare with RNTester's requirement but as far as I have tested production react-native apps on CI with detox and test butler, they work without issues.

So if it's fruitful for us, we can have test butler implemented to avoid android emulator issues like ANRs, system spell checker etc.

File: packages/rn-tester-e2e/README.md
Comment (szymonrybczak) [reply packages/rn-tester-e2e/README.md]: Yeah, sorry for miscommunication. I added instructions to `README.md` how to build with RN CLI using `--buildFolder` flag (which corresponds to `-derivedDatpath`). Btw. I used this in CI, and it works great - saved a lot of time :D

**Comment by szymonrybczak** [comment]
👋 Hello!

Recently I added CircleCI integration, so that on every commit E2E tests are running - and I think we're ready for the final review & merge. 

@NickGerleman @cipolleschi @cortinico could you please look at these changes again? 🙏

**Comment by kelset** [comment]
Can confirm on my end that everything should be ready for final review so that we can get this merged 🤩

File: packages/rn-tester-e2e/README.md
Comment (cipolleschi) [inline packages/rn-tester-e2e/README.md:48]: can we have a js script that we can invoke with like:
```bash
yarn run-e2e-test <platform> # <platform> can be "ios" or "android"
```

That automate the steps required to build and run the tests?
This will make it easier also to ask people to contribute as they can just write their tests and run the script.

File: packages/rn-tester-e2e/tests/helpers/utils.js
Comment (cipolleschi) [inline packages/rn-tester-e2e/tests/helpers/utils.js:41]: where can we find the list of all the supported modifiers? We would need the link to direct contributors that wants to write tests. Is it the Appium docs?

**Comment by cipolleschi** [comment]
Left a couple of extra comments here, but great job guys! :D

File: packages/rn-tester-e2e/README.md
Comment (kelset) [reply packages/rn-tester-e2e/README.md:48]: can we have this being a feature to implement as a separate follow up PR? It doesn't sound like a deal breaker and I'd rather get this in and then have smaller PRs after to build on top of it

File: packages/rn-tester-e2e/tests/helpers/utils.js
Comment (kelset) [reply packages/rn-tester-e2e/tests/helpers/utils.js:41]: I'm not sure what you mean with modifiers here, but for the methods to interact with the elements the list is here -> https://webdriver.io/docs/api/element

the `iOSLabel` and `androidWidget` methods are following the pattern from this section (I believe) https://webdriver.io/docs/selectors/#mobile-selectors

cc @szymonrybczak @mateuszm22 @adzironman in case you have better references

File: packages/rn-tester-e2e/README.md
Comment (szymonrybczak) [reply packages/rn-tester-e2e/README.md:48]: I can handle creating additional PR adding this script, after merging this PR 👍

File: .circleci/config.yml
Comment (cortinico) [inline .circleci/config.yml:811]: Not sure why we're installing `x86_64` but I believe we should be using `x86` here instead (according to CircleCI docs):
https://circleci.com/docs/android-machine-image/

File: .circleci/config.yml
Comment (cortinico) [inline .circleci/config.yml:820]:
```suggestion
          command: npm install appium@2.0.0 -g
```

File: .circleci/config.yml
Comment (cortinico) [inline .circleci/config.yml:841]:
```suggestion
            ./gradlew :packages:rn-tester:android:app:assembleHermesDebug -PreactNativeArchitectures=x86_64
```

or `x86` based on which Arch we pick

File: .circleci/config.yml
Comment (cortinico) [inline .circleci/config.yml:844]: If we change arch, this needs to be updated accordingly

File: packages/rn-tester-e2e/README.md
Comment (cortinico) [inline packages/rn-tester-e2e/README.md:21]:
```suggestion
npm install appium@2.0.0 -g
```

File: .circleci/config.yml
Comment (cortinico) [inline .circleci/config.yml:851]: Nit: maybe this can become a js script?

File: packages/rn-tester-e2e/e2e-config.js
Comment (cortinico) [inline packages/rn-tester-e2e/e2e-config.js:26]: nit: we're scattering `33`, `13.0` all over the codebases, so bumping Android/iOS version would be harder

File: packages/rn-tester-e2e/package.json
Comment (cortinico) [inline packages/rn-tester-e2e/package.json:19]:
```suggestion
    "appium": "^2.0.0",
```

**Review by cortinico — APPROVED** [review]
Great work! 👍 Mostly nits @kelset. This should be good to go also internally, I'm taking a stab at it.

File: .circleci/config.yml
Comment (kelset) [reply .circleci/config.yml:811]: [it fails if we use `x86`](https://app.circleci.com/pipelines/github/facebook/react-native/27861/workflows/895f30e1-7ac5-4701-84f3-4ca778a40965/jobs/859070), we're keeping `x86_64 `

Comment (cortinico) [reply .circleci/config.yml:811]: Oh I see. It's because Google is not distributing system images for `x86` for API level >= 30 anymore. So let's stay with `x86_64` for now.

File: packages/rn-tester-e2e/e2e-config.js
Comment (cortinico) [reply packages/rn-tester-e2e/e2e-config.js:26]: nothing we need to fix here anyway

**Comment by christian-bromann** [comment]
Hey 👋  WebdriverIO maintainer here!

It's excited to see RN using WebdriverIO for their e2e testing. Please don't hesitate to reach out to me if there are any questions or issues coming up. Looking at the code changes I was wondering why you have decided to use Jest as oppose to the WebdriverIO testrunner? The testrunner was designed to allow people to scale their e2e tests and provide good reporting and plugin capabilities. For example it could simplify setting up Appium using the [`@wdio/appium-service`](https://www.npmjs.com/package/@wdio/appium-service).

Any feedback would be greatly appreciated and helps making WebdriverIO better. Cheers!

**Comment by kelset** [comment]
> Looking at the code changes I was wondering why you have decided to use Jest as oppose to the WebdriverIO testrunner?

They @christian-bromann - thanks for your comment! We'll surely reach out if we need help 🤗 About your question, the tl;dr is that using Jest was a strict requirement, you can [read more in the RFC](https://github.com/react-native-community/discussions-and-proposals/pull/684/files#diff-2cd0e7e316a85289b7857c6df8e22e3cd04d5a9168c27fed7720796db8b21af3R46)

**Pull request closed without merging**
